### PR TITLE
Add parquet disk cache for EODHD API responses

### PR DESCRIPTION
**Pull request by michaelchu**
## Summary

- Add `ParquetCache` class for persisting historical options and stock price data as parquet files at `~/.optopsy/cache/{category}/{SYMBOL}.parquet` — no TTL, no eviction (immutable historical data)
- Refactor `EODHDProvider` with a generic `_fetch_with_cache` orchestrator that handles cache read → gap detection → API fetch → merge → slice for both options and stock price endpoints
- Three-tier gap detection: before cached range, interior holes (>5 calendar day threshold), and after cached range — minimizes redundant API calls
- Extract DRY helpers (`_parse_date`, `_coerce_numeric`, `_check_response`, `_safe_raise_for_status`) and split `_apply_options_transforms` into focused functions (`_filter_options`, `_resolve_underlying_prices`, `_select_options_columns`)
- Enable `compact=1` mode on EODHD API for smaller payloads
- Cache stores both calls and puts together; `option_type`/`expiration_type` filtering is client-side only
- Graceful fallback to stale cache when API fails
- 27 new tests covering cache read/write/merge, gap detection edge cases, and orchestrator fallback behavior

## Test plan

- [x] `pytest tests/ -v` — 148 tests pass
- [x] `black --check optopsy/ tests/ setup.py` — clean
- [x] Manual smoke test: fetch options for a ticker, verify cache file created, re-fetch same ticker and confirm no API calls logged
- [x] Verify interior gap re-fetch: partially cache a date range, request a wider range, confirm only gaps are fetched

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -23,6 +23,7 @@
     install_requires=[
         "pandas",
         "numpy",
+        "pyarrow>=14.0.0",
```
[the remark this answers is not in the document]
Comment (michaelchu) [reply setup.py:26]: Wrong. pyarrow 23.0.1 works fine on Python 3.13 — we're running it right now. Copilot's suggested python_version < '3.13' marker would actually break caching on 3.13. Ignored.

File: optopsy/ui/providers/eodhd.py [hunk 2]
```diff
@@ -155,10 +213,150 @@ def _paginate_window(
...
+        overlap_start = max(start_dt, cached_min) if start_dt else cached_min
+        overlap_end = min(end_dt, cached_max) if end_dt else cached_max
+        if overlap_start <= overlap_end:
+            unique_dates = sorted(cached_dates.unique())
+            for prev, curr in zip(unique_dates, unique_dates[1:]):
+                # Only check dates inside the overlap region
+                if prev < overlap_start or curr > overlap_end:
+                    continue
+                day_gap = (curr - prev).days
+                if day_gap > _INTERIOR_GAP_THRESHOLD:
```
[the remark this answers is not in the document]
Comment (michaelchu) [reply optopsy/ui/providers/eodhd.py]: Valid, fixed. The old code skipped gaps where prev < overlap_start or curr > overlap_end, which meant a gap straddling the boundary (e.g. cache has Jan 2 and Mar 1, request is Jan 15–Feb 15) was invisible. Fixed by checking if any gap intersects the overlap instead, and clamping the fetched range to [overlap_start, overlap_end].

File: optopsy/ui/providers/eodhd.py [hunk 3]
```diff
@@ -155,10 +213,150 @@ def _paginate_window(
...
+                        )
+                    )
+
+        # Gap after cached range
+        if end_dt and end_dt > cached_max:
+            gaps.append((str(cached_max + timedelta(days=1)), str(end_dt)))
+        elif end_dt is None:
+            # No end date — fetch from day after cache max onward
+            gaps.append((str(cached_max + timedelta(days=1)), None))
+
```
[the remark this answers is not in the document]
Comment (michaelchu) [reply optopsy/ui/providers/eodhd.py:291]: Valid, fixed by same change. The clamping in the interior gap logic now ensures we never fetch outside the requested range. The before/after gap logic was already correct (before-gap only fires when start_dt < cached_min, after-gap only when end_dt > cached_max).

File: optopsy/ui/providers/eodhd.py [hunk 4]
```diff
@@ -94,59 +147,64 @@ def _get_api_key(self) -> str | None:
...
             if not page_rows:
                 break
 
-            for row in page_rows:
-                attrs = row.get("attributes", row)
-                rows.append(attrs)
+            if fields:
+                # Compact format — zip field names with each row array
+                for row in page_rows:
+                    rows.append(dict(zip(fields, row)))
```
[the remark this answers is not in the document]
Comment (michaelchu) [reply optopsy/ui/providers/eodhd.py:207]: The API either sends compact format or not. Adding a full type-check loop doubles the code for a scenario that won't happen. If the API breaks, we'll know from the exception.

File: optopsy/ui/providers/eodhd.py [hunk 5]
```diff
@@ -319,75 +570,78 @@ def _fetch_options(
...
+                params["to"] = gap_end
+
+            url = f"https://eodhd.com/api/eod/{symbol}.US"
+            resp = self._request_with_retry(url, params)
+            error = _check_response(resp)
+            if error:
+                return error
+            _safe_raise_for_status(resp)
+
+            data = resp.json()
```
[the remark this answers is not in the document]
Comment (michaelchu) [reply optopsy/ui/providers/eodhd.py:642]: Valid concern but extremely unlikely after raise_for_status(). Not worth wrapping every .json() call.

File: optopsy/ui/providers/cache.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+            present = [c for c in dedup_cols if c in merged.columns]
+            if present:
+                merged = merged.drop_duplicates(subset=present, keep="last")
+            else:
+                merged = merged.drop_duplicates()
+        else:
+            merged = merged.drop_duplicates()
+
+        self.write(category, symbol, merged)
+        return merged
```
[the remark this answers is not in the document]
Comment (michaelchu) [reply optopsy/ui/providers/cache.py:80]: The chat UI is single-threaded per session. Documented limitation, not a bug.
